File: Haskell/rpg.hs
-- propriedades em todas as classes
data Comum = Comum { nome :: String
                   , nivel :: Int
                   , experiencia :: Int -- experiencia
                   , dinheiro :: Int -- dinheiro
                   , saude :: Int -- pontos de saude
                   , maxSaude :: Int -- numero maximo de pontos de saude
                   , vel :: Int -- pontos de velocidade do personagem
                   , ataque :: Int -- pontos de ataque do personagem
                   , defesa :: Int -- pontos de defesa do personagem
                   } deriving Show

data Avatar = Guerreiro { padrao   :: Comum 
                        , armadura :: Int -- nivel de armadura
                        , nEspada  :: Int -- nivel da espada
                        , nArco    :: Int -- nivel do arco
                        , nEscudo  :: Int -- nivel de escudo
                        } 
              | 
              Mago { padrao :: Comum
                   , mana :: Int -- pontos de mana
                   , maxMana :: Int -- numero maximo dos pontos de mana
                   , poder :: Int
                   , conhecimento :: Int
                   }
             | 
             Sacerdote { padrao :: Comum
                       , reza :: Int 
                       , maxReza :: Int
                       , cura :: Int
                       , protecao :: Int
                       } 
            deriving Show

                     
                                     
                     

File: rpg-haskell-c/rpg.c
#include<stdio.h>
#include<stdlib.h>

struct Comum{ char nome[50];
              int nivel;
              int experiencia;
              int dinheiro;
              int saude;
              int maxSaude;
              int vel;
              int ataque;
              int defesa;
};
            
enum CLASSE { Guerreiro, Mago, Sacerdote };
struct Avatar { struct Comum padrao;
                enum CLASSE c;
                union {
                    struct { int armadura;
                             int nEspada;
                             int nArco;
                             int nEscudo;
                    };
                    struct { int mana;
                             int maxMana;
                             int poder;
                             int conhecimento;
                    };
                    struct { int reza;
                             int maxReza;
                             int cura;
                             int protecao;
                    };
                };
};

int main(){
	struct GUERREIRO;
	struct Mago;
	struct Sacerdote;
	return 0;
}

